### PR TITLE
Add list of explicitly animatable properties

### PR DESCRIPTION
**Pull request by jverkoey**

### CODE REVIEW

**Comment by jverkoey** [comment]
Ready for review.

**Comment by jverkoey** [comment]
See https://github.com/material-motion/motion-animator-objc/blob/80f2a91aacc5e2b3ab82e5036b6d8dddb15d74b0/README.md for a preview.

File: README.md [hunk 1]
```diff
@@ -7,191 +7,660 @@
 [![CocoaPods Compatible](https://img.shields.io/cocoapods/v/MotionAnimator.svg)](https://cocoapods.org/pods/MotionAnimator)
 [![Platform](https://img.shields.io/cocoapods/p/MotionAnimator.svg)](http://cocoadocs.org/docsets/MotionAnimator)
 
+<table>
+  <tr><td>🎉</td><td>Implicit and explicit additive animations.</td></tr>
+  <tr><td>🎉</td><td>Parameterized motion with the <a href="https://github.com/material-motion/motion-interchange-objc">Interchange</a>.</td></tr>
+  <tr><td>🎉</td><td>Provide velocity to animations directly from gesture recognizers.</td></tr>
+  <tr><td>🎉</td><td>Minimize jank by relying more on Core Animation.</td></tr>
```
Comment (romoore) [inline README.md]: Can we come up with a more precise term than "jank"? What exactly are we minimizing? Is it measurable?

Comment (jverkoey) [reply README.md]: Done.

**Comment by romoore** [comment]
Sorry, I think this is just too long for me to sit and review all at once.

**Comment by jverkoey** [comment]
Ah that's fair - I can split it up into separate pieces.

**Comment by jverkoey** [comment]
This PR is ready for re-review. It now includes only the animatable key paths charts and additions of the quiz tests to the unit tests.

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 2]
```diff
@@ -29,52 +29,101 @@ import MotionAnimator
...
+  func testViewDoesImplicitlyAnimateInUIViewAnimateBlock() {
+    let view = UIView()
+    UIView.animate(withDuration: CATransaction.animationDuration() + 0.1) {
+      view.alpha = 0.5
+    }
+
+    XCTAssert(view.layer.animation(forKey: "opacity") is CABasicAnimation)
+    if let animation = view.layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, CATransaction.animationDuration() + 0.1)
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit: Please use XCTAssertEqualWithAccuracy

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 3]
```diff
@@ -29,52 +29,101 @@ import MotionAnimator
...
+  func testViewLayerDoesImplicitlyAnimateInUIViewAnimateBlock() {
+    let view = UIView()
+    UIView.animate(withDuration: CATransaction.animationDuration() + 0.1) {
+      view.layer.opacity = 0.5
+    }
+
+    XCTAssert(view.layer.animation(forKey: "opacity") is CABasicAnimation)
+    if let animation = view.layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, CATransaction.animationDuration() + 0.1)
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit: Please use XCTAssertEqualWithAccuracy

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 4]
```diff
@@ -29,52 +29,101 @@ import MotionAnimator
...
     layer.opacity = 0.5
-    CATransaction.commit()
 
-    let animation = layer.animation(forKey: "opacity") as! CABasicAnimation
-    XCTAssertEqual(animation.keyPath, "opacity")
-    XCTAssertEqual(animation.duration, 0.5)
+    XCTAssert(layer.animation(forKey: "opacity") is CABasicAnimation)
+    if let animation = layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, CATransaction.animationDuration())
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit: Please use XCTAssertEqualWithAccuracy

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 5]
```diff
@@ -29,52 +29,101 @@ import MotionAnimator
...
     unflushedLayer.opacity = 0.5
-    CATransaction.commit()
 
     XCTAssertNil(unflushedLayer.animationKeys())
   }
 
-  func testDoesImplicitlyAnimateInCATransaction() {
-    CATransaction.begin()
-    CATransaction.setAnimationDuration(0.5)
+  func testDoesImplicitlyAnimate() {
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit/Naming: "**What** is implicitly animating?"  `testUnhostedLayerInWindowDoesImplicitlyAnimate`?

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 6]
```diff
@@ -29,52 +29,101 @@ import MotionAnimator
...
+    window.layer.addSublayer(layer)
+    CATransaction.flush()
+    UIView.animate(withDuration: 0.8, animations: {
+      layer.opacity = 0.5
+    })
+
+    XCTAssert(layer.animation(forKey: "opacity") is CABasicAnimation)
+    if let animation = layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, CATransaction.animationDuration())
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit: Please use XCTAssertEqualWithAccuracy

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 7]
```diff
@@ -83,67 +132,125 @@ class HeadlessLayerImplicitAnimationTests: XCTestCase {
   }
 
   func testCATransactionTimingTakesPrecedenceOverUIViewTimingInside() {
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: This looks like the same as `testImplicitlyAnimatesInUIViewAnimateBlockWithCATransactionDuration`. Am I missing something?

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 8]
```diff
@@ -83,67 +132,125 @@ class HeadlessLayerImplicitAnimationTests: XCTestCase {
...
+    if let animation = layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, CATransaction.animationDuration())
+    }
   }
 
   // Verifies the somewhat counter-intuitive fact that CATransaction's animation duration always
   // takes precedence over UIView's animation duration. This means that animating a headless layer
   // using UIView animation APIs may not result in the expected traits.
   func testCATransactionTimingTakesPrecedenceOverUIViewTimingOutside() {
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Again, I must be missing a difference in this test. It looks identical to `testImplicitlyAnimatesInUIViewAnimateBlockWithCATransactionDuration`

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 9]
```diff
@@ -83,67 +132,125 @@ class HeadlessLayerImplicitAnimationTests: XCTestCase {
...
-    XCTAssertEqual(animation.keyPath, "opacity")
-    XCTAssertEqual(animation.duration, 0.2)
+    XCTAssert(layer.animation(forKey: "opacity") is CABasicAnimation)
+    if let animation = layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, CATransaction.animationDuration())
+    }
   }
 
   func testDoesImplicitlyAnimateInUIViewAnimateBlock() {
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Maybe comments will help to distinguish these cases? Looks just like `testImplicitlyAnimatesInUIViewAnimateBlockWithCATransactionDuration` to me.

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 10]
```diff
@@ -83,67 +132,125 @@ class HeadlessLayerImplicitAnimationTests: XCTestCase {
...
+      layer.opacity = 0.5
     }
 
-    let animation = layer.animation(forKey: "opacity") as! CABasicAnimation
-    XCTAssertEqual(animation.keyPath, "opacity")
-    XCTAssertEqual(animation.duration, 0.2)
+    XCTAssert(layer.animation(forKey: "opacity") is CABasicAnimation)
+    if let animation = layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, CATransaction.animationDuration())
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit: Please use XCTAssertEqualWithAccuracy

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 11]
```diff
@@ -83,67 +132,125 @@ class HeadlessLayerImplicitAnimationTests: XCTestCase {
...
     }
-    CATransaction.commit()
 
-    let animation = layer.animation(forKey: "opacity") as! CABasicAnimation
-    XCTAssertEqual(animation.keyPath, "opacity")
-    XCTAssertEqual(animation.duration, 0.2)
+    XCTAssert(layer.animation(forKey: "opacity") is CABasicAnimation)
+    if let animation = layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, CATransaction.animationDuration())
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit: Please use XCTAssertEqualWithAccuracy

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 12]
```diff
@@ -83,67 +132,125 @@ class HeadlessLayerImplicitAnimationTests: XCTestCase {
...
+    CATransaction.flush()
+
+    UIView.animate(withDuration: CATransaction.animationDuration() + 0.1) {
+      layer.opacity = 0.5
+    }
+
+    XCTAssert(layer.animation(forKey: "opacity") is CABasicAnimation)
+    if let animation = layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, CATransaction.animationDuration())
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit: Please use XCTAssertEqualWithAccuracy

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 13]
```diff
@@ -83,67 +132,125 @@ class HeadlessLayerImplicitAnimationTests: XCTestCase {
...
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit: Please use XCTAssertEqualWithAccuracy

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 14]
```diff
@@ -83,67 +132,125 @@ class HeadlessLayerImplicitAnimationTests: XCTestCase {
...
 
-    let animation = layer.animation(forKey: "opacity") as! CABasicAnimation
-    XCTAssertEqual(animation.keyPath, "opacity")
-    XCTAssertEqual(animation.duration, CATransaction.animationDuration())
+    layer.opacity = 0.5
+
+    XCTAssert(layer.animation(forKey: "opacity") is CABasicAnimation)
+    if let animation = layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, CATransaction.animationDuration())
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit: Please use XCTAssertEqualWithAccuracy

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift [hunk 15]
```diff
@@ -152,19 +259,25 @@ class HeadlessLayerImplicitAnimationTests: XCTestCase {
...
     }
     CATransaction.commit()
 
-    let animation = layer.animation(forKey: "opacity") as! CABasicAnimation
-    XCTAssertEqual(animation.keyPath, "opacity")
-    XCTAssertEqual(animation.duration, traits.duration)
+    XCTAssert(layer.animation(forKey: "opacity") is CABasicAnimation)
+    if let animation = layer.animation(forKey: "opacity") as? CABasicAnimation {
+      XCTAssertEqual(animation.keyPath, "opacity")
+      XCTAssertEqual(animation.duration, traits.duration)
```
Comment (romoore) [inline tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Nit: Please use XCTAssertEqualWithAccuracy

**Review by romoore — CHANGES_REQUESTED** [review]

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift (hunk 7 above)
Comment (jverkoey) [reply tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Removed.

File: tests/unit/HeadlessLayerImplicitAnimationTests.swift (hunk 8 above)
Comment (jverkoey) [reply tests/unit/HeadlessLayerImplicitAnimationTests.swift]: Removed.

**Pull request merged**